featurebyte/app.py

```diff
@@ -126,6 +128,75 @@ async def _wrapper(
     return _wrapper
 
 
+ws_connections: set[WebSocket] = set()
+
+
+async def close_ws_connections() -> None:
+    """
+    Close all websocket connections.
+    """
+    for ws in ws_connections:
+        await ws.close(code=status.WS_1001_GOING_AWAY)
+
+
+async def redis_to_websocket(websocket: WebSocket, channel: str) -> None:
+    """
+    Forward messages from Redis channel to Websocket
+
+    Parameters
+    ----------
+    websocket: WebSocket
+        Websocket object
+    channel: str
+        Channel name
+    """
+    async with redis.from_url(REDIS_URI) as client:
+        async with client.pubsub() as pubsub:
+            logger.debug("Listening to channel", extra={"channel": channel})
+            await pubsub.subscribe(channel)
+
+            # Listen to the channel and forward the messages to the websocket
+            try:
+                async for message in pubsub.listen():
+                    if message and isinstance(message, dict):
+                        data = message.get("data")
+                        if isinstance(data, bytes):
+                            await websocket.send_bytes(data)
+            # Client disconnected
+            except WebSocketDisconnect as e:
+                logger.debug("Websocket disconnected", extra={"exception": e})
+                if websocket.client_state == WebSocketState.CONNECTED:
+                    await websocket.close(code=status.WS_1000_NORMAL_CLOSURE)
+            except Exception as e:
+                logger.error("Websocket connection error", extra={"exception": e})
+                # Close client connection if still active
+                if websocket.client_state == WebSocketState.CONNECTED:
+                    await websocket.close(code=status.WS_1011_INTERNAL_ERROR)
+            finally:
+                ws_connections.remove(websocket)
```

Comment:
WIll doing the close in finally not get triggered on shutdown?

Reply:
We need to add `global ws_connections` here to use the variable.

Reply:
Will put the global in. 

Im not sure the order of operations for the server shutdown, whether the lifecycle stop fires first; closing the connection, or the finally block gets triggered by an exception.
Im just putting both sides as a safeguard.

Reply:
@kchua78  I've looked at the global part. 
I'm not mutating the ws_connections pointer, so it should work without adding the `global`.